src/python/pants/core/goals/check.py

```diff
@@ -201,8 +203,33 @@ async def check(
         )
         for request_type in request_types
     )
+
+    environment_names = await MultiGet(
+        Get(
+            EnvironmentName,
+            EnvironmentNameRequest,
+            EnvironmentNameRequest.from_field_set(field_set),
+        )
+        for request in requests
+        for field_set in request.field_sets
+    )
+    request_for_each_environment_name = (
+        # this is actually correct:
+        request
+        for request in requests
+        for _ in request.field_sets
+    )
+    grouped = itertools.groupby(
+        zip(request_for_each_environment_name, environment_names), lambda x: x[0]
```

Comment:
Rather than zipping, consider emitting both the request and environment name in `request_for_each_environment_name`?